File: backend/shared/uidl/src/main/java/io/mateu/uidl/data/FormField.java
package io.mateu.uidl.data;

import io.mateu.uidl.fluent.Component;
import java.util.List;
import java.util.Map;
import lombok.Builder;

@Builder(toBuilder = true)
public record FormField(
    String id,
    String label,
    FieldDataType dataType,
    FieldStereotype stereotype,
    boolean readOnly,
    boolean required,
    boolean autofocus,
    String placeholder,
    String description,
    List<Option> options,
    RemoteCoordinates remoteCoordinates,
    RestDataSource optionsSource,
    Object initialValue,
    String style,
    String cssClasses,
    List<GridContent> columns,
    boolean inlineEditing,
    String onItemSelectionActionId,
    String rowSelectionShortcut,
    int colspan,
    FormPosition formPosition,
    String formStyle,
    String formTheme,
    int formColumns,
    int sliderMin,
    int sliderMax,
    boolean stepButtonsVisible,
    double step,
    Double min,
    Double max,
    String itemIdPath,
    Map<String, String> attributes,
    String detailPath,
    boolean useButtonForDetail,
    String minHeightWhenDetailVisible,
    int optionsColumns,
    boolean mainFilter,
    // tree selects (stereotype treeSelect): restrict selection to leaf nodes
    boolean treeLeavesOnly,
    boolean multiline,
    // property-list sections (@Section(propertyList=true)): render as a read-only row with the
    // label aligned left and the plain-text value aligned right, separated by a divider line
    boolean propertyRow,
    NavLink link)
    implements Component {

  public FormField {
    stereotype = stereotype != null ? stereotype : FieldStereotype.regular;
  }

  @Override
  // A YAML-authored field declares only what it needs, so these two arrive null on a `filters:`
  // entry that says nothing about them — and every mapper reads them by `.name()`. GridColumn
  // already defends the same two; FormField did not, so a filter without a `dataType` answered the
  // whole screen with a NullPointerException.
  public FieldDataType dataType() {
    return dataType != null ? dataType : FieldDataType.string;
  }

  public FieldStereotype stereotype() {
    return stereotype != null ? stereotype : FieldStereotype.regular;
  }

  public List<Option> options() {
    return options != null ? options : List.of();
  }

  public List<GridContent> columns() {
    return columns != null ? columns : List.of();
  }
}

File: frontend/web/monorepo/libs/mateu/src/mateu/shared/apiClients/dtos/componentmetadata/FormField.ts
import Validation from "./Validation";
import Badge from "./Badge";
import Option from "@mateu/shared/apiClients/dtos/componentmetadata/Option";
import ComponentMetadata from "@mateu/shared/apiClients/dtos/ComponentMetadata";
import ClientSideComponent from "@mateu/shared/apiClients/dtos/ClientSideComponent.ts";
import RemoteCoordinates from "@mateu/shared/apiClients/dtos/componentmetadata/RemoteCoordinates.ts";
import NavLink from "@mateu/shared/apiClients/dtos/componentmetadata/NavLink.ts";

export default interface FormField extends ComponentMetadata {

    fieldId: string
    dataType: string
    stereotype: string
    observed?: boolean | undefined
    wantsFocus?: boolean | undefined
    label: string | undefined
    placeholder?: string | undefined
    cssClasses?: string | undefined
    description?: string | undefined
    badges?: Badge[] | undefined
    validations?: Validation[] | undefined
    colspan?: number | undefined
    rightAligned?: boolean | undefined
    bold?: boolean | undefined
    initialValue?: unknown
    options?: Option[] | undefined
    remoteCoordinates: RemoteCoordinates | undefined
    charLimit?: string
    required: boolean
    disabled: boolean
    readOnly: boolean
    style?: string | undefined
    columns: ClientSideComponent[]
    createForm: ClientSideComponent
    editor: ClientSideComponent
    onItemSelectionActionId: string | undefined
    rowSelectionShortcut: string | undefined
    inlineEditing: boolean
    formPosition: string
    formStyle: string
    formTheme: string
    formColumns: number
    sliderMin: number
    sliderMax: number
    stepButtonsVisible: boolean
    step: number
    min?: number | undefined
    max?: number | undefined
    itemIdPath: string | undefined
    detailPath: string
    useButtonForDetail: boolean
    minHeightWhenDetailVisible: string
    optionsColumns: number
    mainFilter?: boolean
    multiline?: boolean
    link?: NavLink | undefined

}
